tests/lib/external/snapd-testing-tools/utils/log-filter

```diff
@@ -0,0 +1,224 @@
+
+        rule_operation = parts[0]
+        rule_pattern = parts[1]
+
+        if rule_operation not in SUPPORTED_RULES:
+            raise ValueError("Error: Rule operation '{}' not in supported list: {}".format(rule_operation, SUPPORTED_RULES))
+
+        if operation == rule_operation:
+            patterns.append(rule_pattern)
+
+    for pattern in patterns:
+        regex_list.append(re.compile(pattern))
+    return regex_list
+
+
+def process_detail_line(line: str, regex_list: list, to_stream: io.TextIOWrapper):
+        if not regex_list:
+            write_line(line, to_stream)
+        else:
+            for regex in regex_list:
+                matches = regex.findall(line)
+                for match in matches:
+                    write_line(match, to_stream)
+
+def process_detail(from_stream: io.TextIOWrapper, start_line: str, regex_list: list, to_stream: io.TextIOWrapper) -> str:
+    """Process lines from the start of the details section until the last line,
+    returns the first line right after the details section
+    """
+    write_line(start_line, to_stream)
+    for line in sys.stdin:
+        print_line(line)
+
+        if not line:
+            continue
+
+        # Check if the detail is finished
+        if is_detail_finished(line):
+            return line
+
+        # Print all the lines
+        process_detail_line(line, regex_list,  to_stream)
+
+            
+def skip_detail(from_stream: io.TextIOWrapper) -> str:
+    """Skip lines from the start of the details section until the last line,
+    returns the first line right after the details section
+    """
+    for line in sys.stdin:
+        print_line(line)
+        if not line:
+            continue
+
+        # Check if the detail is finished
+        if is_detail_finished(line):
+            return line
+
+def process_spread_output(output_file, exclude_lines, filter_rules):
+    error_regex = compile_rules(filter_rules, ERROR)
+    debug_regex = compile_rules(filter_rules, DEBUG)
+    failed_regex = compile_rules(filter_rules, FAILED)
+    warning_regex = compile_rules(filter_rules, WARNING)
+
+    with open(output_file, "w") as myfile:
+        for line in sys.stdin:
+            print_line(line)
+
+            while is_detail_start(line):
+                regex_list = []
+                if is_line(line, DEBUG):
+                    if DEBUG in exclude_lines:
+                        line = skip_detail(sys.stdin)
+                        continue
+                    else:
+                        regex_list = debug_regex
+                
+                if is_line(line, ERROR):
+                    if ERROR in exclude_lines:
+                        line = skip_detail(sys.stdin)
+                        continue
+                    else:
+                        regex_list = error_regex
+
+                if is_line(line, WARNING):
+                    if WARNING in exclude_lines:
+                        line = skip_detail(sys.stdin)
```

Comment:
hm wasn't the exclusion supposed to filter out the lines from stdout and write them to a file? isn't this operating in reverese?

Reply:
well, those tools are for different things, this tool is needed to filter the output to send to the observability stack, the other tool is used to filter the output displayed in github page. 
So, it will work like this -> spread ... | filter_1 | filter_2
The filter 1 will save to a new file the filtered output and will send to stdout the stdin
The filter 2 will save to files the debug output and will send to stdout the contect without the debug output